accounts-db/benches/lock_accounts.rs

```diff
@@ -0,0 +1,151 @@
+}
+
+fn create_funded_accounts(bank: &Bank, num: usize) -> Vec<Keypair> {
+    assert!(
+        num.is_power_of_two(),
+        "must be power of 2 for parallel funding tree"
+    );
+    let accounts = create_accounts(num);
+
+    accounts.par_iter().for_each(|account| {
+        bank.store_account(
+            &account.pubkey(),
+            &Account {
+                lamports: 5100,
+                data: vec![],
+                owner: system_program::id(),
+                executable: false,
+                rent_epoch: Epoch::MAX,
+            }
+            .to_account_shared_data(),
+        );
+    });
+
+    accounts
+}
+
+fn create_transactions(bank: &Bank, num: usize) -> Vec<SanitizedTransaction> {
+    let funded_accounts = create_funded_accounts(bank, 2 * num);
+    funded_accounts
+        .into_par_iter()
+        .chunks(2)
+        .map(|chunk| {
+            let from = &chunk[0];
+            let to = &chunk[1];
+            system_transaction::transfer(from, &to.pubkey(), 1, bank.last_blockhash())
+        })
+        .map(SanitizedTransaction::from_transaction_for_tests)
+        .collect()
+}
+
+fn bank_setup() -> Arc<Bank> {
+    let mint_total = u64::MAX;
+    let GenesisConfigInfo {
+        mut genesis_config, ..
+    } = create_genesis_config(mint_total);
+
+    // Set a high ticks_per_slot so we don't run out of ticks
+    // during the benchmark
+    genesis_config.ticks_per_slot = 10_000;
+
+    let mut bank = Bank::new_for_benches(&genesis_config);
+
+    // Allow arbitrary transaction processing time for the purposes of this bench
+    bank.ns_per_slot = u128::MAX;
+
+    // set cost tracker limits to MAX so it will not filter out TXs
+    bank.write_cost_tracker()
+        .unwrap()
+        .set_limits(u64::MAX, u64::MAX, u64::MAX);
+    bank.wrap_with_bank_forks_for_tests().0
+}
+
+fn bench_lock_accounts(
+    bencher: &mut Bencher,
+    batch_size: usize,
+    allow_self_conflicting_entries: bool,
+) {
+    const TRANSACTIONS_PER_ITERATION: usize = 64;
+    assert_eq!(
+        TRANSACTIONS_PER_ITERATION % batch_size,
+        0,
+        "batch_size must be a factor of `TRANSACTIONS_PER_ITERATION` \
+         ({TRANSACTIONS_PER_ITERATION}) so that bench results are easily comparable"
+    );
+    let batches_per_iteration = TRANSACTIONS_PER_ITERATION / batch_size;
+
+    let bank = bank_setup();
+    let transactions = create_transactions(&bank, 2_usize.pow(20));
+    let batches: Vec<_> = transactions.chunks(batch_size).collect();
+
+    bencher.iter(|| {
+        for i in 0..batches_per_iteration {
+            if let Some(batch) = batches.get(i) {
+                let _ = bank.rc.accounts.lock_accounts(
+                    (*batch).iter(),
+                    MAX_TX_ACCOUNT_LOCKS,
+                    allow_self_conflicting_entries,
+                );
+            }
+        }
+    });
+
+    // drop batches here so dropping is not included in the benchmark
+    drop(batches);
```

Comment:
Need to unlock, otherwise we just fail fast - for me this made bench times >3x longer on master. Haven't compared against the changes in PR yet.

Also probably don't need so many txs, we can just cycle iterator and use black_box to make compiler pessimistic about what is inside the batch.

```suggestion
    let transactions = create_transactions(&bank, 2_usize.pow(16));
    let mut batches = transactions.chunks(batch_size).cycle();
    bencher.iter(|| {
        for batch in (0..batches_per_iteration).filter_map(|_| batches.next()) {
            let results = bank
                .rc
                .accounts
                .lock_accounts(test::black_box(batch.iter()), MAX_TX_ACCOUNT_LOCKS);
            bank.rc.accounts.unlock_accounts(batch.iter().zip(&results));
        }
    });
```